src/hooks/useResizeObserver/useResizeObserver.ts

```diff
@@ -1,24 +1,44 @@
+import { debounce } from 'lodash-es';
 import { type RefObject, useEffect } from 'react';
 
+const resizeObserverInstanceCallbacks = new Map<Element, ResizeObserverCallback>();
+let resizeObserverInstance: ResizeObserver | null = null;
+
 /**
  * This hook allows you to add a ResizeObserver for an element and remove it
  * when the component unmounts.
  *
  * @param ref - The ref to observe
  * @param callback - The callback to fire when the element resizes
+ * @param debounceTime - The number in milliseconds the callback is debounced
  */
-export function useResizeObserver(ref: RefObject<Element>, callback: ResizeObserverCallback): void {
+export function useResizeObserver(
+  ref: RefObject<Element>,
+  callback: ResizeObserverCallback,
+  debounceTime?: number | null,
+): void {
   useEffect(() => {
-    const resizeObserverInstance = new ResizeObserver(callback);
-
     if (ref.current === null) {
       throw new Error('`ref.current` is undefined');
     }
 
-    resizeObserverInstance.observe(ref.current);
+    if (!resizeObserverInstance) {
+      resizeObserverInstance = new ResizeObserver((entries, observer) => {
```

Comment:
I think the previous approach is fine, so I'd like to keep the implementation as simple as possible. Yeah, the old implementation is creating multiple resize observers which may not be necessary, but from a performance point of view it's not likely that we're creating thousands of resize observers on a single page.